[Debugger] Added support for stepping over await and out of async methods

DavidKarlas opened the pull request:
PR only comment:

* This PR should be reviewed per commit:
  * 1st commit [ppdb] is basically implementing more ppdb metadata reading https://github.com/dotnet/roslyn/blob/2ae8d5fed96ab3f1164031f9b4ac827f53289159/docs/specs/PortablePdb-Metadata.md and Deserialization of AsyncDebugInfo from https://github.com/dotnet/roslyn/blob/c65f9cc0723ecf52ff15445f38f45c0bc44aa482/src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs#L559
  * 2nd commit is actual async stepping.
* Since runtime is not my home... Please check for runtime(e.g. handling MonoError correctly) and C mistakes(e.g. casting and freeing stuff).
* Is there better way to compare hardcoded GUID then `is_async_method_stepping_information_guid`?
* @vargaz is `get_this` method robust/safe enough?

This is new PR because I made formatting changes and rebased on master and forced pushed, but looks like PR doesn't detect changes on force push... I addressed bug I found and @vargaz commets: https://github.com/mono/mono/pull/4206#issuecomment-270675499

vargaz commented:
Please use the mono runtime coding conventions.

vargaz commented on mono/metadata/debug-mono-ppdb.c:
```diff
@@ -595,3 +596,130 @@ mono_ppdb_lookup_locals (MonoDebugMethodInfo *minfo)
+static gboolean
+is_async_method_stepping_information_guid (guint8* debug_info_kind_guid)
+{
+	return debug_info_kind_guid [0] == 0xC5 &&
```
Would be better to declare this as a static guint8 array and compare using a loop.

vargaz replied:
Where is this guid from ?

vargaz commented on mono/metadata/debug-mono-ppdb.c:
```diff
@@ -595,3 +596,130 @@ mono_ppdb_lookup_locals (MonoDebugMethodInfo *minfo)
+		debug_info_kind_guid [15] == 0xF8;
+}
+
+//for parent_type see HasCustomDebugInformation table at https://github.com/dotnet/roslyn/blob/2ae8d5fed96ab3f1164031f9b4ac827f53289159/docs/specs/PortablePdb-Metadata.md#customdebuginformation-table-0x37
```
Break this line of comment.

vargaz replied:
the url is this:
https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/specs/PortablePdb-Metadata.md

vargaz commented on mono/metadata/debug-mono-ppdb.c:
```diff
@@ -595,3 +596,130 @@ mono_ppdb_lookup_locals (MonoDebugMethodInfo *minfo)
+	int blob_len = mono_metadata_decode_blob_size (blob, &blob);
+	MonoDebugMethodAsyncInfo* res = g_new0 (MonoDebugMethodAsyncInfo, 1);
+	char const *pointer = blob;
+	pointer += 4;//catch_handler_offset
```
Where is this documented ?

DavidKarlas replied:
https://github.com/dotnet/roslyn/blob/1ad4b58/src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs#L566

vargaz replied:
Can this be added as a comment.

vargaz commented on mono/metadata/mono-debug.h:
```diff
@@ -186,6 +187,9 @@ mono_debug_add_delegate_trampoline (void* code, int size);
 mono_debug_lookup_locals (MonoMethod *method);
 
+MONO_API MonoDebugMethodAsyncInfo*
+mono_debug_lookup_method_async_debug_info (MonoMethod *method);
```
This shouldn't be MONO_API until embedders need it.

DavidKarlas commented:
@vargaz Added new commit addressing all comments, except not sure what `Please use the mono runtime coding conventions.` is referring to, since I thought I addressed all such problems.

P.S: I can later squash this commit.

vargaz commented on mono/metadata/mono-debug-debugger.h:
```diff
@@ -20,7 +20,7 @@ mono_debug_get_seq_points (MonoDebugMethodInfo *minfo, char **source_file, GPtrA
 MONO_API void
 mono_debug_free_locals (MonoDebugLocalsInfo *info);
 
-MONO_API void
```
This should probably be moved to mono-debug.h, and we should get rid of mono-debug-debugger.h, although this is a different issue than this PR.

DavidKarlas replied:
I was just following `mono_debug_free_locals`, hence added MONO_API in first place...

vargaz replied:
Its ok.

vargaz commented:
The second commit still doesn't conform to mono coding conventions, e.g. variable/function naming, brace placement, etc. Runtime code doesn't use camelCase.

DavidKarlas commented:
Fixed formatting problems

DavidKarlas commented:
@vargaz not sure you noticed, I think I fixed all problems now

vargaz commented on mono/mini/debugger-agent.c:
```diff
@@ -4608,6 +4646,91 @@ breakpoint_matches_assembly (MonoBreakpoint *bp, MonoAssembly *assembly)
+		return;
+	MonoClassField *builderField = mono_class_get_field_from_name (obj->vtable->klass, "<>t__builder");
+	if (builderField == NULL) {
+		DEBUG_PRINTF (1, "[%p] Failed to get <>t__builder.\n", (gpointer)(gsize)mono_native_thread_id_get ());
```
Most of these checks can be asserts, they shouldn't fail.

vargaz commented on mono/mini/debugger-agent.c:
```diff
@@ -4608,6 +4646,91 @@ breakpoint_matches_assembly (MonoBreakpoint *bp, MonoAssembly *assembly)
+		mono_defaults.corlib, "System.Threading.Tasks", "Task");
+	GPtrArray* array = mono_class_get_methods_by_name (taskClass, "NotifyDebuggerOfWaitCompletion", 0x24, FALSE, FALSE, &error);
+	if (!is_ok (&error)) {
+		DEBUG_PRINTF (1, "[%p] Failed to get NotifyDebuggerOfWaitCompletion method: %s.\n", (gpointer)(gsize)mono_native_thread_id_get (), mono_error_get_message (&error));
```
Thse should be asserts.

vargaz commented on mono/mini/debugger-agent.c:
```diff
@@ -4608,6 +4646,91 @@ breakpoint_matches_assembly (MonoBreakpoint *bp, MonoAssembly *assembly)
+static MonoMethod*
+get_notify_debugger_of_wait_completion_method ()
+{
+	MonoError error;
```
The result should be cached in a static var.

vargaz commented on mono/mini/debugger-agent.c:
```diff
@@ -4696,10 +4819,45 @@ process_breakpoint_inner (DebuggerTlsData *tls, gboolean from_signal)
+
+			//Check method is async before calling get_this_async_id
+			MonoDebugMethodAsyncInfo* asyncMethod = mono_debug_lookup_method_async_debug_info (method);
+			if (!asyncMethod)
```
All these calls to mono_debug_lookup_method_async_debug_info () look kinda expensive.

DavidKarlas replied:
This is only called when async stepping is in progress I think it shouldn't be hot path... Is there cheaper way to check if method is async or should we cache result?

vargaz commented on mono/mini/debugger-agent.c:
```diff
@@ -4563,19 +4591,29 @@ ss_update (SingleStepReq *req, MonoJitInfo *ji, SeqPoint *sp, DebuggerTlsData *t
 	}
 
+	MonoDebugMethodAsyncInfo* asyncMethod = mono_debug_lookup_method_async_debug_info (method);
+	if (asyncMethod) {
```
This looks kinda expensive too.
Also, we do we add those seq points in method-to-ir.c if we ignore them here ?

DavidKarlas replied:
We add them in method-to-ir.c mostly because otherwise we can't place breakpoints(we don't have ManagedIL->naitve_offset relation AFAICT), but also so when we are stepping we step on top of yield_offset.

I was thinking about if I should cache `mono_debug_lookup_method_async_debug_info` calls... I figured `ss_start` isn't called often(imho should only be called when user request stepping)...

vargaz replied:
Why doesn't the debug infor contain seqpoint points at these locations, should the c# compiler do that ?

vargaz replied:
Also, I still think most of this work should be in the ide, not in the runtime. I'm pretty sure thats how .net works.

DavidKarlas replied:
C# compiler doesn't emit sequence points for this locations because this offsets are special and are emitted as special sequence points.
We can always move stepping logic for from runtime to IDE, but for now stepping logic is in runtime.

kumpera replied:
Then this is the perfect time to move it. This is overly complicated code to have in the runtime debugger bits.

DavidKarlas commented:
@vargaz I replaced debug logging with asserting and added caching of MonoMethod fetching.

I didn't cache/optimise mono_debug_lookup_method_async_debug_info calls, since as I said it shouldn't be hot path.
I think we should see if this is problem in future and optimise if it's problem.

DavidKarlas commented:
Fixing asserting...

DavidKarlas commented:
@vargaz pushed much cleaner asserts

vargaz commented:
Looks ok, please squash and merge.

DavidKarlas commented:
Since I squashed into 2 commits I did it manually and pushed to master

kumpera commented on mono/metadata/debug-mono-ppdb.c:
```diff
@@ -595,3 +596,126 @@ mono_ppdb_lookup_locals (MonoDebugMethodInfo *minfo)
+} locator_t;
+
+static int
+table_locator (const void *a, const void *b)
```
This is duplicate code from metadata

kumpera commented on mono/metadata/debug-mono-ppdb.c:
```diff
@@ -595,3 +596,126 @@ mono_ppdb_lookup_locals (MonoDebugMethodInfo *minfo)
+
+static gboolean
+compare_guid (guint8* guid1, guint8* guid2) {
+	for (int i = 0; i < 16; i++) {
```
Use memcmp

kumpera commented on mono/mini/debugger-agent.c:
```diff
@@ -4526,18 +4530,42 @@ static void ss_calculate_framecount (DebuggerTlsData *tls, MonoContext *ctx)
+ensure_jit (StackFrame* frame)
+{
+	if (!frame->jit) {
+		frame->jit = mono_debug_find_method (frame->api_method, frame->domain);
```
This looks wrong, why can't it be checked upfront for a ginst method?

kumpera commented on mono/mini/debugger-agent.c:
```diff
@@ -5191,6 +5342,28 @@ ss_bp_add_one (SingleStepReq *ss_req, int *ss_req_bp_count, GHashTable **ss_req_
+	mono_seq_point_init_next (info, *sp, next);
+	for (int i = 0; i < sp->next_len; i++) {
+		if (next [i].flags & MONO_SEQ_POINT_FLAG_NONEMPTY_STACK) {
+			if (!is_last_non_empty (&next [i], info)) {
```
This is recursive. How deep can it go?
